runtime/dart_isolate.cc

```diff
@@ -0,0 +1,615 @@
+                                            package_config,              //
+                                            flags,                       //
+                                            callback_data,               //
+                                            error                        //
+    );
+  }
+
+  return CreateDartVMAndEmbedderObjectPair(
+             advisory_script_uri, advisory_script_entrypoint, package_root,
+             package_config, flags, callback_data, error)
+      .first;
+}
+
+std::pair<Dart_Isolate, fxl::WeakPtr<DartIsolate>>
+DartIsolate::CreateDartVMAndEmbedderObjectPair(
+    const char* advisory_script_uri,
+    const char* advisory_script_entrypoint,
+    const char* package_root,
+    const char* package_config,
+    Dart_IsolateFlags* flags,
+    void* callback_data,
+    char** error) {
+  // The callback data that was passed to the parent isolate.
+  auto parent_embedder_isolate = static_cast<const DartIsolate*>(callback_data);
+
+  if (parent_embedder_isolate == nullptr ||
+      parent_embedder_isolate->GetDartVM() == nullptr) {
+    *error =
+        strdup("Parent isolate did not have embedder specific callback data.");
+    return {nullptr, {}};
+  }
+
+  const DartVM* vm = parent_embedder_isolate->GetDartVM();
+
+  // Create the native object on the embedder side. This object is deleted in
+  // the cleanup callback.
+  auto embedder_isolate =
+      new DartIsolate(vm, parent_embedder_isolate->GetThreads(),
+                      advisory_script_uri, advisory_script_entrypoint);
+
+  // Create the Dart VM isolate and give it the embedder object as the baton.
+  Dart_Isolate isolate =
+      vm->GetPlatformKernel() != nullptr
+          ? Dart_CreateIsolateFromKernel(advisory_script_uri,         //
+                                         advisory_script_entrypoint,  //
+                                         vm->GetPlatformKernel(),     //
+                                         nullptr,                     //
+                                         embedder_isolate,            //
+                                         error                        //
+                                         )
+          : Dart_CreateIsolate(
+                advisory_script_uri,                                    //
+                advisory_script_entrypoint,                             //
+                vm->GetDartSnapshot().GetDefaultIsolateSnapshotData(),  //
+                vm->GetDartSnapshot()
+                    .GetDefaultIsolateSnapshotInstructions(),  //
+                nullptr,                                       //
+                embedder_isolate,                              //
+                error                                          //
+            );
+
+  if (!embedder_isolate->Initialize(isolate)) {
+    *error = strdup("Embedder could not initialize the Dart isolate.");
+    return {nullptr, {}};
```

Comment:
Does Dart_ShutdownIsolate() need to be called here (and in the other error case below)?

Reply:
I don't think so. If an isolate is created and nothing run in it, the VM should shut down the isolate on its own since the engine holds no strong reference to the same. The runtime `DartIsolate` created by the embedder will then be collected in the cleanup callback below.

Reply:
Two things:
- An Isolate needs to be able to survive it's window/drawing surface coming and going.
- We need to take an isolate that has already been asked to run one entrypoint, e.g. main(), and tell it to run another entrypoint, e.g. timerCallback().